Monitor QPRDC and run jit flush if threshold exceeded

dpino opened the pull request:
Monitors QPRDC register (Queue Packet Received Drop Counter), and calls `jit.flush()` if packet drop threshold is exceeded.

Original code by Marcel Wiget. Back ported from [1].

[1] https://github.com/mwiget/snabbswitch/commit/89cec535790f8d5762d31824e55582169a2bcfe7

wingo commented on src/apps/intel/intel10g.lua:
```diff
@@ -19,13 +19,16 @@ local index_set = require("lib.index_set")
 num_descriptors = 512
 --num_descriptors = 32
 
+local ifInDiscards = counter.open("nic/ifInDiscards")
```
Probably not the right thing for us, given that we need to instantiate multiple NICs.  Marcel can get away with just this one because he runs the lwAFTR on a stick.

wingo commented:
OK here is a stupid idea.  Right, Snabb already monitors statistics for all the links.  But it doesn't monitor drop statistics on ingress (or egress for that matter) -- and that's where drops often happen.  What if, for apps which have a pull() method, we add some kind of standard method to get the number of dropped packets, corresponding to ifInDiscards.  Snabb monitors that and exposes it as a counter.  Then we can decide to jit.flush() in core Snabb.  It would seem to be a workaround that should apply to everyone.  WDYT?  If you think the idea is terrible, fine :)  Otherwise maybe we draft such a thing and send it upstream.

mwiget commented:
@wingo: makes perfect sense to me, because it allows a logic to be added to core that takes multiple parameters into account prior to triggering jit.flush(). My latest version uses a threshold knob for lost packets plus a grace period before a new flush() can get triggered. It takes a few seconds to stabilize.
This will also allow multiple apps/interfaces with external, but measurable ingress buffer problems to signal such issue to core Snabb.

dpino commented:
OK, it took me a while to understand the proposed change. I think I got it. The idea is for packet-generator apps (apps with a pull method) to monitor their ingress packet drop in a counter (following the same pattern as for the links). Then in Snabb's main loop the counter for each pulling app is queried and a jit.flush() is called in case the counter has exceeded a certain limit.

@mwiget What would be the best wait of setting the packet drop threshold. I'm thinking of defining a global in `core/app.lua` with a default value. Programs can modify this variable as they need it. OTOH, I think I backported the patch from your `dual` branch. If there's a more updated branch I can take a look at it. I can add a grace period before flush() variable too.

mwiget commented:
@dpino The commit with the wait knob is https://github.com/mwiget/snabbswitch/commit/fd709f2152cf74ab0ac1a4fcde68797cbc8025c4

Its in branch 1to1_mapping. The 'dual' branch is an experimental one, where I drive two VhostUsers interface with a single 10GE port. That simplified the logic in snabb (no v4v6 splitter), but didn't really bring any performance gains just yet (and not really needed for the single stick design, but potentially other use cases).

In respect to default value and allow a user to overwrite it, yes, as a parameter to engine.main() makes sense to me.

mwiget commented:
The wait knob was a quick fix to fine tune the effect, because just using a threshold per second required me to set it to 100k. The naming of that knob can probably also be improved and even the algorithm.

dpino committed:
Monitor ingress packet drops and jit flush if threshold exceeded

dpino commented:
Updated PR.

In the end I didn't need to use counters. Since the engine needs to check whether a pull app has a method to retrieve the number of ingress dropped packets, writing this value to a counter is not necessary. The engine can call the app's method directly.

Still, if we want to write the value to a counter that would be possible. Let me know.

wingo commented on src/apps/intel/intel10g.lua:
```diff
@@ -187,6 +192,21 @@ do
+end
+
+function M_sf:ingress_packet_drops ()
+   return self.qprdc
```
what would be the perf impact if we just read self.qs.QPRDC[0] here, and avoided the timer?

wingo replied:
also does reading the qprdc reset it to zero or something?  otherwise not sure how SF configurations would reset.  (edit: never mind, apparently the comment below indicates that reading it resets the value)

dpino replied:
The reason why resetting the value to zero was that the value of QPRDC is stored in self.qprdc but it only gets updated every second. The engine polls the value several times per second so in the case it needs to jit.flush it will do it continuously until a new QPRDC value from the card is read, that's the reason to set it to zero once read.

It slipped my mind to do the same for SF configurations.

If we skip the timer and read QPRDC continuously resetting to zero after a read won't be necessary. I will measure perf of this PR and without the timer. Let's see how it goes.

wingo replied:
Let's accumulate the QPRDC value into a counter, when we read it, and return a continuously increasing counter.  Seems to be safest :)

mwiget commented:
@dpino hmm, there is actually tremendous value in making the counter available to an external app like top to document if and how many packets get dropped at ingress. This has been invaluable in understanding where exactly packets are dropped in a real network test scenario under line rate load.
Any chance or idea where best to expose those? Ideally, they would be great to have in 'snabb top', but obviously per actual 10GE port.

mwiget commented:
@wingo: hmm, if reading that register resets its value, I'm worried about the impact when someone also enables snmp reporting, which also queries that register for ifInDiscards.

dpino commented:
@mwiget So it seems it's worth to write this value to a counter. I had some doubts on the tree structure of the counter. When I first read Andy's suggestion what it came to my mind was that make each pull app have an `ingress_packet_drops` counter, so the tree would go like this `/pullapps/<app_name>/ingress_packet_drops`. Is that OK?

There are some challenges with creating this tree structure (apps don't know their name until `pull` is called) but I think I can find a solution.

Exposing the values in snabb top sounds good to me. I would leave that for another PR.

So, after this feeback I will update the PR to implement the following changes:

* Write QPRDC to a counter continuously (without a timer).
* Make the engine read a pullapp counter and jit.flush if packet drop exceeded.

How does it sound?

wingo commented:
dpino: dunno, ideally naming-wise this meshes with https://github.com/snabbco/snabb/pull/766; not sure though.  good luck on names, we will bikeshed it to death :)  anyway what you say sounds good to me!

mwiget commented:
@dpino: I like your proposal. Using` <app_name> ` in the tree works for single and multiple ports to get reported individually. I make the assumption here that the `<app_name>` is the string set in `config_app(c, app_name, ...)`.  Its also human readable, at least once one passes the hurdle of figuring out which process id to look at ;-).

dpino commented:
@mwiget Yep, <app_name> will be the name of the app that is passed to config.app.

@wingo Thanks for the pointer, I can use `/app` as root instead of `/pullapp`. At first glimpse, the issue that I see with that PR is the shm path would only be available when the links and apps are setup (after `engine.configure()` has been called). In that case the app only has the chance to create its counters on `pull` (what I commented before). I think the shm path needs to be setup on `config.app` and passed as an argument to the app, so the app can create its counters once in `app:new()` https://github.com/snabbco/snabb/blob/master/src/core/config.lua#L27

dpino commented:
Updated PR. PTAL.

wingo commented on src/apps/intel/intel10g.lua:
```diff
@@ -187,6 +187,10 @@ do
 end
 
+function M_sf:ingress_packet_drops ()
+   return tonumber(self.qs.QPRDC[0]())
```
Let's leave it as a uint64.  I know it will usually be small, but usually tonumber() should only be applied to counter diffs; most places to store counter values are uint64 so they can hold the whole value.

I still don't quite understand the bit about resetting to zero.  AFAIU this register is an "RC" register (see the end of the file), and lib.hardware.register will map that to "register.readrc", which keeps its own accumulator.  Which means we don't have to mess about with counters, even; SNMP should work fine AFAIU.  If that's true then the app doesn't need to register a counter (though it can), and we can do that in some separate PR.  Sound about right to you?

dpino commented:
OK. I have split the PR in two. This one reads `ingress_packet_drops` and jit.flush if threshold is exceeded. There's a grace period between flushes.

The other PR is #303. It exposes the `ingress_packet_drops` in a counter. Each app has its own counter. The app depends on https://github.com/snabbco/snabb/pull/905 or whatever other solution that will eventually be implemented to allow apps to obtain their own name and do not wait to create their counters until `pull` starts.

wingo commented on src/apps/intel/intel_app.lua:
```diff
@@ -35,7 +35,7 @@ end
 function Intel82599:new (arg)
    local conf = config.parse_app_arg(arg)
-
+   local result
```
this change seems unnecessary or an artifact of previous things; please remove:)

wingo commented on src/core/app.lua:
```diff
@@ -283,18 +288,35 @@ function pace_breathing ()
 
+local function exceeded_ingress_packet_drops (app)
+   if app.ingress_packet_drops then
+      return app:ingress_packet_drops() > ingress_packet_drops.threshold
```
given that ingress_packet_drops() should be monotonically increasing (modulo uint64 wraparound), this means that once we exceed the limit we will always exceed the limit.  probably not the right thing!

wingo commented:
looking good, but i think we probably need to keep track of the last counter value at which we flushed.

dpino commented:
Updated PR. PTAL.

wingo commented:
Hi, this still isn't looking right to me -- resetting the counter each time you call the function means that probably we never reach the threshold.  Am I missing something?  If not, then please fix by making ingress_packet_drops() return a monotonically increasing value, and store the value at which the last flush occurred in core/app.lua.  Thanks!

dpino commented:
@wingo sorry, I misunderstood how it should work. I think is fine now. Let me know what do you think.

wingo commented on src/core/app.lua:
```diff
@@ -20,6 +20,12 @@ local use_restart = false
+ingress_packet_drops = {
+   threshold = 100000,
+   wait = 20,
+   last_value = 0,
```
In order to avoid allocations from the uint64 return from `:ingress_packet_drops()` from escaping, probably this should be `last_value = ffi.new('uint64_t[0]')`, and accesses should go through `ingress_packet_drops.last_value[0].`

wingo replied:
but, maybe this is for a followup, we can release on monday like this; dunno.  we'll need to do performance testing too, and that will take a bit of time.

wingo commented on src/core/app.lua:
```diff
@@ -283,18 +289,41 @@ function pace_breathing ()
+      jit.flush()
+      last_trigger = now + ingress_packet_drops.wait
+   end
+end)()
```
I would go ahead and smash the stateful logic together into one function, where we return true if we should call jit.flush(), taking into account both the threshold and the wait parameter.  (Incidentally we will probably want to expose the number of jit.flush() calls in a counter, in the future, and maybe the time of the last flush; pretty important operational data.)  WDYT?  Maybe these are bad ideas :)

wingo commented:
Looking really great, just a nit.

dpino committed:
Smash jit flush logic into one single function

dpino commented:
OK, SGTM. In the end what I did was to smash the logic into one single function `jit_flush_if_needed` that also calls `jit.flush`. I added a new patch instead of amending. Let me know if that's what you had in mind.

I like the idea of expose number of jit.flush and flush time in counters. I can do that in a next PR once this one lands.

wingo commented on src/core/app.lua:
```diff
@@ -283,18 +289,35 @@ function pace_breathing ()
       if app.pull and not app.dead then
          zone(app.zone)
          with_restart(app, app.pull)
+         jit_flush_if_needed(app)
```
Diego I am so sorry to nitpick this thing to death, but I have another significant comment.

We're currently keeping a global counter for the "number of dropped packets the last time that we did a flush" value (`ingress_packet_drops.last_value`), but there can be more than one app, and their packet drop counts don't necessarily correlate.

Either we keep a per-app baseline for drops, or we sum drops across all apps.  I suggest we do the latter.  What do you think?  I suggest doing

``` lua
local ingress_drop_monitor = {
   threshold = 100000,
   wait = 20,
   last_flush = 0,
   last_value = ffi.new('uint64_t[1]'),
   current_value = ffi.new('uint64_t[1]')
}

function ingress_drop_monitor:sample()
   local sum = self.current_value
   sum[0] = 0
   for i = 1, #app_array do
      local app = app_array[i]
      if app.pull and not app.dead then
         sum[0] = sum[0] + app:ingress_packet_drops()
      end
   end
end

function ingress_drop_monitor:jit_flush_if_needed()
   if self.current_value[0] - self.last_value[0] < self.threshold then return end
   if app.now() - self.last_flush < self.wait then return end
   self.last_flush = app.now()
   self.last_value[0] = self.current_value[0]
   jit.flush()
   --- in the future we change the last_flush, last_value, and current_value fields to be counters
end
``
```

wingo replied:
Finally we need to measure the overhead of this solution.  If we are processing 4e6 PPS and each breath pulls in 1e2 packets, that's 40000 breaths/second; probably more than needed.  I suggest 100ms as the sampling period.  WDYT?  We can then just add a timer to the engine and make the timer call `sample()`, which I guess could include the jit_flush_if_needed call; dunno.  Then we add a parameter to app.run() that controls whether we attach the timer or not.  WDYT?  I dunno, maybe they are bad ideas, I am just getting nervous about upstream's willingness to include such a thing in the core loop :)

dpino commented:
@wingo Everything sounds very reasonable. I'm OK with these changes. Making this feature optional would likely increase its chances to be merged upstream. I will update the PR.

dpino committed:
Wrap ingress drop monitor logic in its own object

- Make ingress drop monitor an engine.main option.
- A timer constantly monitors ingress drops and jit flush
if threshold is exceeded.

dpino commented:
Updated PR. PTAL.

wingo commented on src/core/app.lua:
```diff
@@ -246,6 +275,16 @@ function main (options)
    end
 
+   if options.ingress_drop_monitor then
+      local interval = 1e9 / 1e2   -- Every 100 milliseconds.
```
This interval is every 10 ms, no?  1ms is 1e6 nanoseconds, so every 100ms would be 1e8 ns.

wingo commented on src/core/app.lua:
```diff
@@ -27,6 +27,35 @@ link_table, link_array = {}, {}
+   sum[0] = 0
+   for i = 1, #app_array do
+      local app = app_array[i]
+      if app.pull and not app.dead then
```
probably need `and app.ingress_packet_drops`

wingo commented on src/core/app.lua:
```diff
@@ -246,6 +275,16 @@ function main (options)
       breathe = latency:wrap_thunk(breathe, now)
    end
 
+   if options.ingress_drop_monitor then
```
Let's default to on :)  i.e. `if options.ingress_drop_monitor or options.ingress_drop_monitor == nil then`, as above with the latency measurer.

wingo commented on src/core/app.lua:
```diff
@@ -27,6 +27,35 @@ link_table, link_array = {}, {}
+   if app.now() - self.last_flush < self.wait then return end
+   self.last_flush = app.now()
+   self.last_value[0] = self.current_value[0]
+   jit.flush()
```
Let's log or something when we do a jit.flush().  A print at the least.

dpino committed:
Fix last commit

dpino commented:
Applied last fixes.

wingo commented:
lgtm, let's pr upstream to see what snabbot says.  thank you!
